Add ability to generate exit message

valefar-on-discord opened the pull request:
Port from https://github.com/ethereum/staking-deposit-cli/pull/355

Adds two commands:
**exit-transaction-keystore** and **exit-transaction-mnemonic**

### exit-transaction-keystore 

Allows the user to specify a single keystore file to generate an exit message if the provided password is correct. 
The success of this command will result in a single `signed_exit_transaction-***.json` file in `exit_transactions`

### exit-transaction-mnemonic

Allows the user to specify a mnemonic to generate 1-n exit messages depending on how many validator indices are provided.  The success of this command will result in n `signed_exit_transaction-***.json` files in `exit_transactions` for each validator index provided.

### Testing
Build as instructed

Create a new mnemonic and keystore(s) on mainnet with
```./deposit.sh new-mnemonic``` 

For each keystore, create an entry in this base `offline-preparation.json` file:
```
{
  "version":"3",
  "validators":[
    {
      "index": "<ANY_INDEX_IE_123456>",
      "pubkey": "0x<PUBKEY>",
      "state": "active_ongoing",
      "withdrawal_credentials": "<CREDENTIALS>"
    }
  ],
  "genesis_validators_root":"0x4b363db94e286120d76eb905340fdd4e54bfe9f06bf33ff6cf5ad27f511bfe95",
  "epoch":"274294",
  "genesis_fork_version":"0x00000000",
  "exit_fork_version":"0x03000000",
  "current_fork_version":"0x04000000",
  "bls_to_execution_change_domain_type":"0x0a000000",
  "voluntary_exit_domain_type":"0x04000000"
}
```

With the saved `offline-preparation.json` above in the same location as ethdo, create an exit message with ethdo using the following command:
```./ethdo validator exit --offline --mnemonic="<GENERATED_MNEMONIC>" --validator=<VALIDATOR_INDEX_IN_PREF_FILE>```

Create an exit message with the CLI using both
```./deposit.sh exit-transaction-mnemonic --epoch=274294```
for each keystore
```./deposit.sh exit-transaction-keystore --epoch=274294```

All outputs should match

remyroy commented on tests/test_cli/test_exit_transaction_keystore.py:
```diff
@@ -0,0 +1,81 @@
+)
+
+
+def test_exit_transaction_menmonic() -> None:
```
We are only testing the happy and successful path for exit-transaction-keystore. It would be nice to add a few tests for the paths where it fail.

valefar-on-discord replied:
Also the name is wrong :D

Going to add a couple of edge cases here.

One case that I don't think is possible to test is chain mismatch. If you create a validator on holesky and use that keystore to generate an exit message with mainnet specified, it will create one, though invalid.  If you know of a way to verify the chain, let me know and I'll add logic for it!

remyroy replied:
From my understanding, the signed exit message depends on the `fork_version` and `genesis_validators_root` for the domain. We could test this by generating a different exit message for all the of testnets and testing that it fails with the `verify_signed_exit_json`/`validate_signed_exit` functions by passing the mainnet chain_settings for instance. That would test the chain mismatch.

remyroy replied:
My bad. I think you cannot know for which chain a keystore was made for so that part cannot be tested, ie. a validator deposited on holekey, using that validator keystore and creating an exit message for mainnet. That exit message should always fail when broadcasted.

remyroy commented on tests/test_cli/test_exit_transaction_mnemonic.py:
```diff
@@ -0,0 +1,99 @@
+from tests.test_cli.helpers import clean_exit_transaction_folder, read_json_file, verify_file_permission
+
+
+def test_exit_transaction_menmonic() -> None:
```
We are only testing the happy and successful path for exit-transaction-mnemonic. It would be nice to add a few tests for the paths where it fail.

remyroy commented on tests/test_cli/test_exit_transaction_mnemonic.py:
```diff
@@ -0,0 +1,99 @@
+    # Verify file content
+    assert json_data['message']['epoch'] == '1234'
+    assert json_data['message']['validator_index'] == '1'
+    assert json_data['signature']
```
In addition to checking for the existence of the signature, it should test that the BLS signature is valid.

valefar-on-discord replied:
see above

remyroy commented on tests/test_cli/test_exit_transaction_mnemonic.py:
```diff
@@ -0,0 +1,99 @@
+        json_data = read_json_file(exit_transaction_folder_path, exit_transaction_files[index])
+        assert json_data['message']['epoch'] == '1234'
+        assert json_data['message']['validator_index'] == str(index)
+        assert json_data['signature']
```
In addition to checking for the existence of the signature, it should test that the BLS signature is valid.

valefar-on-discord replied:
I'm going to add test cases for the specific validation logic but that will be tested with an exit code of 0.

remyroy reviewed (changes requested):
Everything is generally good. I still have some real world tests I would like to complete on Holesky. Let me know what you think of these comments.

valefar-on-discord commented on staking_deposit/cli/existing_mnemonic.py:
```diff
@@ -47,7 +47,7 @@ def load_mnemonic_arguments_decorator(function: Callable[..., Any]) -> Callable[
             help=lambda: load_text(['arg_mnemonic_password', 'help'], func='existing_mnemonic'),
             hidden=True,
-            param_decls='--mnemonic-password',
+            param_decls='--mnemonic_password',
```
This was a reported issue: https://github.com/ethereum/staking-deposit-cli/issues/353

The documentation uses `_` but the param was `-`

The convention looks to be have command `-` separated and arguments `_` separated.

remyroy replied:
That seems good. Thanks for the clarification.

valefar-on-discord commented on staking_deposit/cli/exit_transaction_keystore.py:
```diff
@@ -0,0 +1,117 @@
+    try:
+        secret_bytes = saved_keystore.decrypt(keystore_password)
+    except ValueError:
+        click.echo(load_text(['arg_exit_transaction_keystore_keystore_password', 'mismatch']))
```
I've updated to
`Error: The provided keystore password was unable to decrypt this keystore file. Make sure you have the correct password and try again.`

If we have a message saying "Did you ..." they might type "yes" :)

remyroy replied:
An even better message. 👍

valefar-on-discord commented on staking_deposit/cli/exit_transaction_keystore.py:
```diff
@@ -0,0 +1,117 @@
+        epoch: int,
+        output_folder: str,
+        **kwargs: Any) -> None:
+    saved_keystore = Keystore.from_file(keystore)
```
I've updated the logic to verify the validity of the keystore file during the input flow.

Sadly, this showed a bit of an issue with the validation logic:

There is a function in `click.py` `captive_prompt_callback` which traps the user in the input loop until the validation function succeeds. The problem here is the Click Option `type` param which can be used to get basic validation of the input is not utilized in this loop.

The result is if you enter the prompt loop, you get a different error message for FileNotFound

A bit annoying but I think the UX of looping until a proper keystore file is found out-weighs the difference in error handling.

remyroy replied:
I'm not a big fan of trapping the user if the file format is wrong. Trapping the user if the file does not exist seems ok. It's not a big deal and I think this could be resolved somewhere else. I'll mark this conversation as resolved for now.

valefar-on-discord commented on staking_deposit/cli/exit_transaction_keystore.py:
```diff
@@ -0,0 +1,117 @@
+        epoch=epoch,
+    )
+
+    saved_folder = export_exit_transaction_json(folder=output_folder, signed_exit=signed_exit)
```
Updated the logic to include a verification flow.

I've moved some of the logic around but wasn't able to completely isolate it as the logic between keystore and mnemonic is enough to be annoying.

Also added a progress bar to the mnemonic flow similar to key generation

remyroy replied:
Good job on this. This works fine on the `exit-transaction-keystore` command. I love the progress bar on the `exit-transaction-mnemonic` command even if that is somewhat unimportant in the grand scheme of things.

valefar-on-discord commented on staking_deposit/cli/exit_transaction_mnemonic.py:
```diff
@@ -0,0 +1,115 @@
+
+        saved_folder = export_exit_transaction_json(folder=output_folder, signed_exit=signed_voluntary_exit)
+        click.echo(load_text(['msg_creation_success']) + saved_folder)
+
```
ack

remyroy replied:
This was completed.

valefar-on-discord commented on staking_deposit/cli/exit_transaction_mnemonic.py:
```diff
@@ -0,0 +1,115 @@
+    click.echo(load_text(['msg_creation_start']))
+    # We assume that the list of validator indices are in order and increment the start index
+    for key_index, validator_index in zip(key_indices, validator_indices):
+        credential = Credential(
```
This can also occur when using the existing-mnemonic flow. 

There are a number of locations where a ValidationError is thrown that will result in the stack trace and error message.

I think wrapped try/catches might be a bit much and perhaps we should remove the ValidationError from more expected locations such as an invalid mnemonic. 

Though if we do want to go that route, we might want to do that in a separate pull request.

What do you think?

remyroy replied:
I agree. I created #18 to keep track of this.

valefar-on-discord commented on tests/test_cli/helpers.py:
```diff
@@ -49,3 +56,8 @@ def prepare_testing_folder(os_ref, testing_folder_name='TESTING_TEMP_FOLDER'):
+
+
+def read_json_file(path: str, file_name: str):
+    with open(os.path.join(path, file_name), 'r') as f:
```
Added explicit encoding when reading a json file

valefar-on-discord commented on tests/test_cli/test_exit_transaction_keystore.py:
```diff
@@ -0,0 +1,81 @@
+    # Verify file content
+    assert json_data['message']['epoch'] == '1234'
+    assert json_data['message']['validator_index'] == '1'
+    assert json_data['signature']
```
This will be done for free in the generation logic

valefar-on-discord commented:
Ran through all the items and added quite a few changes:
- Added logic to verify the integrity of the exit signature for both keystore and mnemonic flow
- Added familiar progress bar with exit mnemonic flow
- Updated some wording for error messages
- Added `NON_PRATER_CHAIN_KEYS` to reduce chain param logic duplication
- Added utf-8 encoding when reading json file
- Removing int cast in `validate_validator_indices` so the value will be validated in `validate_int_range` instead where it will throw the proper error

If I missed anything, let me know and thanks for such a thorough review

remyroy commented on tests/test_utils/test_validation.py:
```diff
@@ -60,3 +62,38 @@ def test_validate_int_range(num: Any, low: int, high: int, valid: bool) -> None:
+invalid_signature = (
+    "0x8cceb99d17361031e01dfb6aa997554a35f60bcc8a106ac76fdea6f5a4780fb8b65b4cd827bca0c88b340508b69f577"
+    "50122db94c319aa05a0165e71b41f30c0982c415727b7e2387cce78d995acd54f038b743dc1426b0fb0d4783617d4fe6e"
+)
```
I think it would be better to generate all these values (valid_pubkey, valid_signature, invalid_pubkey, invalid_signature) on the fly, during each test, with our current code from original mnemonics, chain settings, etc instead of using pre-baked hex data for which we don't know much about. Not only will it make the test more readable and easier to understand, it will likely increase the overall test coverage (testing more of the pubkey and signature generation part or retesting it again).

valefar-on-discord replied:
This might be a difference of opinion with testing.
My idea was we are testing only `validate_signed_exit` and we should be explicit with our inputs and then have expected results similar to some tests in `test_ssz.py` and `test_crypto.py`.
If we expand this test scenario to utilize mnemonic, key, and exit generation, then I would consider that an integration test and should belong elsewhere. 

If we generated the logic on the fly, we would end up with something like for the setup
```
    mainnet_chain_settings = get_chain_setting("mainnet")
    holesky_chain_settings = get_chain_setting("holesky")
    mnemonic = get_mnemonic(language="en", words_path=WORD_LISTS_PATH)
    mnemonic2 = get_mnemonic(language="en", words_path=WORD_LISTS_PATH)
    correct_keystore= Credential(
        mnemonic=mnemonic,
        mnemonic_password="",
        index=0,
        amount=0,
        chain_setting=mainnet_chain_settings,
        hex_eth1_withdrawal_address=None,
    )
    wrong_chain_keystore = Credential(
        mnemonic=mnemonic,
        mnemonic_password="",
        index=0,
        amount=0,
        chain_setting=holesky_chain_settings,
        hex_eth1_withdrawal_address=None,
    )
    wrong_mnemonic_keystore = Credential(
        mnemonic=mnemonic2,
        mnemonic_password="",
        index=0,
        amount=0,
        chain_setting=mainnet_chain_settings,
        hex_eth1_withdrawal_address=None,
    )
    correct_exit = correct_keystore.generate_exit_transaction(0, 0)
    wrong_index_exit = correct_keystore.generate_exit_transaction(1, 0)
    wrong_epoch_exit = correct_keystore.generate_exit_transaction(0, 1)
    wrong_chain_exit = wrong_chain_keystore.generate_exit_transaction(0, 0)
    wrong_mnemonic_exit = wrong_mnemonic_keystore.generate_exit_transaction(0, 0)
```

Which is quite a bit.

How about we keep this as is and I create a separate integration test suite that does this dynamic generation so we can sweep through more of the code base?
Or if you feel unit tests are not an effective use of time for CI, I can replace the unit testing logic with something similar to the above.

remyroy replied:
That is a good point. Let's leave it as is for now and see how to do other tests elsewhere.

remyroy reviewed (changes requested):
I think there are 1 minor thing remaining before I can approve this:

- Replacing the pre-baked hex data in `tests/test_utils/test_validation.py`

I'm still waiting on my validators to be activated on Holeksy to test these changes on a real network. According to [beaconcha.in](https://holesky.beaconcha.in/validator/a5d9dfdf8ce7e0c4ea9010a0e0bc8203d799d5efe45b58fec87b37d81e1e6095cb288fa4337cf544e418486f59134eab#deposits), I still have around 6 days to wait.

remyroy commented:
> > We should add the `exit_transactions` directory, the default directory for exit transaction, in the `.gitignore` file similar to how we have `validator_keys` and `bls_to_execution_changes` to avoid potentially polluting the repo.
> 
> I thought I did that. Did I mess something up there?

Sorry, I missed it with being on the wrong branch. Everything is good here.

remyroy reviewed (approved):
Everything is good in the review. What's left are some real world tests that I should be able to perform in the next few days.

remyroy commented:
My validators were just activated on Holesky and I tried to submit the exit messages but I have to wait another 2 days until I can start the exit process.

remyroy commented:
I've successfully exited 2 validators, one with a keystore and one with a mnemonic.

1. https://holesky.beaconcha.in/validator/1688054
2. https://holesky.beaconcha.in/validator/1688057

remyroy commented:
I tried to resolve the pending merge conflict but I think I broke the tests. It feels like they are stuck waiting for some input on stdin.

valefar-on-discord commented:
There was a problem with the git merge that combined two changes in the existing mnemonic test where it provided the argument for the execution address and input.
So the test stalled expecting verification of the execution address argument which it never got.
